### PR TITLE
Implement the pointer event overhaul

### PR DESCRIPTION
**Pull request by daxpedda**
### Description

This implements #3833, please see it for all the details.

This implementation deviates on `PointerLeft::position`, which is now an `Option`, because Windows (and maybe MacOS) doesn't have this kind of information.
It also adds `FingerId` to `PointerType`, which was an oversight of #3833.

Notably I also removed pen/stylus handling in iOS and Windows, which is now emitted as `PointerType::Unknown`.
I plan to reintroduce the same code and expand on it in #3810.

#3874 was a side-product of this PR.

### Summary of the changes
- Rename `CursorMoved` to `PointerMoved`.
- Rename `CursorEntered` to `PointerEntered`.
- Rename `CursorLeft` to `PointerLeft`.
- Rename `MouseInput` to `PointerButton`.
- Add `position` to every `PointerEvent`.
- Remove `Touch`, which is folded into the `Pointer*` events.
- New `PointerType` added to `PointerEntered` and `PointerLeft`, signifying which pointer type is the source of this event.
- New `PointerSource` added to `PointerMoved`, similar to `PointerType` but holding additional data.
- New `ButtonSource` added to `PointerButton`, similar to `PointerType` but holding pointer type specific buttons. Use `ButtonSource::mouse_button()` to easily normalize any pointer button type to a generic mouse button.
- In the same spirit rename `DeviceEvent::MouseMotion` to `PointerMotion`.
- Remove `Force::Calibrated::altitude_angle`.

### Follow-ups
- Implement pen/stylus input in #3810.
- Add touch pressure support to X11.
- Add a reliable way to detect pointer types to X11.
- Potentially differentiate between `TouchPad` and `TouchScreen`. `Touch` will remain in place for backends who can not differentiate between these pointer types.

---

Addresses https://github.com/rust-windowing/winit/issues/1555.
Addresses https://github.com/rust-windowing/winit/issues/99.
Fixes https://github.com/rust-windowing/winit/issues/336.
Fixes https://github.com/rust-windowing/winit/issues/883.
Fixes https://github.com/rust-windowing/winit/issues/1909.
Replaces #3001.

### CODE REVIEW

File: src/platform_impl/apple/appkit/view.rs [hunk 1]
```diff
@@ -639,19 +639,32 @@ declare_class!(
...
 
         #[method(mouseEntered:)]
-        fn mouse_entered(&self, _event: &NSEvent) {
+        fn mouse_entered(&self, event: &NSEvent) {
             trace_scope!("mouseEntered:");
-            self.queue_event(WindowEvent::CursorEntered {
+
+            let Some(position) = self.mouse_view_point(event) else {
+                return
+            };
```
Comment (daxpedda) [inline src/platform_impl/apple/appkit/view.rs:647]: @madsmtm I couldn't find in the [documentation](https://developer.apple.com/documentation/appkit/nsevent/1529068-locationinwindow) if this works.

Same for `mouseExited`.

Comment (madsmtm) [reply src/platform_impl/apple/appkit/view.rs:647]: The documentation says that it works for all "mouse events":
> This property applies to mouse events

This includes `mouseEntered:` and `mouseExited:` (I've tested that it works, and indeed it does).

File: src/event.rs [hunk 2]
```diff
@@ -222,51 +222,86 @@ pub enum WindowEvent {
...
-    /// - **Web:** Doesn't take into account CSS [`border`], [`padding`], or [`transform`].
-    ///
-    /// [`border`]: https://developer.mozilla.org/en-US/docs/Web/CSS/border
-    /// [`padding`]: https://developer.mozilla.org/en-US/docs/Web/CSS/padding
-    /// [`transform`]: https://developer.mozilla.org/en-US/docs/Web/CSS/transform
-    CursorLeft { device_id: DeviceId },
+    PointerLeft {
+        device_id: DeviceId,
+
+        /// The position of the cursor when it left the window.
```
Comment (madsmtm) [inline src/event.rs:268]: This definition is troublesome: If the cursor has left the window, then it will be outside the window's bounds, and the position might be non-sensical for the current application.

I see a few options (in order of my preference):
- Define that pointer positions in general can be negative and outside the bounds of the window (on some platforms).
- "Clamp" the position to be within the bounds of the window.
- Define this event to be "the position the cursor had just before it left the window", i.e. store pointer events, and re-emit the position from the last `PointerMoved`.

File: src/event.rs [hunk 3]
```diff
@@ -425,6 +448,118 @@ pub enum WindowEvent {
...
+    /// [`WindowEvent::PointerLeft`] event is generated with the same [`FingerId`].
+    ///
+    /// After a [`WindowEvent::PointerEntered`] event has been emitted, there may be zero or more
+    /// [`WindowEvent::PointerMoved`] events when the finger is moved or the touch pressure
+    /// changes.
+    ///
+    /// A [`WindowEvent::PointerLeft`] without a [`WindowEvent::PointerButton`] with
+    /// [`ElementState::Released`] event is emitted when the system has canceled tracking this
+    /// touch, such as when the window loses focus, or on mobile devices if the user moves the
+    /// device against their face.
```
Comment (madsmtm) [inline src/event.rs:493]: This behaviour of getting `Pressed` without `Released` when a touch phase is cancelled, is IMO quite subtle. Perhaps we should rather expose `PointerCancel`?

File: src/event.rs [hunk 4]
```diff
@@ -425,6 +448,118 @@ pub enum WindowEvent {
...
 }
 
+/// Represents the pointer type of a pointer event.
+///
+/// ## Platform-specific
+///
+/// **Wayland/X11:** Can only identify [`Touch`](Self::Touch), any other pointer is identified as
+/// [`Mouse`](Self::Mouse).
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum PointerType {
```
Comment (madsmtm) [inline src/event.rs:458]: Splitting this from `PointerSource` does not feel sufficiently motivated, but perhaps it will make sense in the future because we're going to get more data on `PointerSource`?

If the issue is the size of `WindowEvent`, I'd argue we `Box` the `Force` struct instead.

File: src/event.rs [hunk 5]
```diff
@@ -832,50 +967,6 @@ pub enum TouchPhase {
...
-    pub location: PhysicalPosition<f64>,
-    /// Describes how hard the screen was pressed. May be `None` if the platform
-    /// does not support pressure sensitivity.
-    ///
-    /// ## Platform-specific
-    ///
-    /// - Only available on **iOS** 9.0+, **Windows** 8+, **Web**, and **Android**.
-    /// - **Android**: This will never be [None]. If the device doesn't support pressure
-    ///   sensitivity, force will either be 0.0 or 1.0. Also see the
-    ///   [android documentation](https://developer.android.com/reference/android/view/MotionEvent#AXIS_PRESSURE).
```
Comment (madsmtm) [inline src/event.rs:873]: Nit: Re-use this documentation in `PointerSource::Touch.force` (or write something else there)

File: src/event.rs (hunk 4 above)
Comment (madsmtm) [reply src/event.rs:458]: If we do keep it, then a naming suggestion could be to use `kind` and `PointerKind` instead of `ty` and `PointerType`.

File: src/event.rs [hunk 6]
```diff
@@ -222,51 +222,86 @@ pub enum WindowEvent {
...
+        ///
+        /// - **Orbital: Always emits `(0., 0.)`.
+        /// - **Web:** Doesn't take into account CSS [`border`], [`padding`], or [`transform`].
+        ///
+        /// [`border`]: https://developer.mozilla.org/en-US/docs/Web/CSS/border
+        /// [`padding`]: https://developer.mozilla.org/en-US/docs/Web/CSS/padding
+        /// [`transform`]: https://developer.mozilla.org/en-US/docs/Web/CSS/transform
+        position: PhysicalPosition<f64>,
+
+        button: ButtonSource,
```
Comment (madsmtm) [inline src/event.rs:303]: I _feel_ like I'd rather mirror web even closer here, and have:
```suggestion
        button: Option<MouseButton>,
        source: PointerSource,
```

Instead of creating a whole `ButtonSource` which ends up containing much of the same information. But that may be an oversimplication, and perhaps it doesn't align well with future plans for pointer events?

**Review by madsmtm — CHANGES_REQUESTED** [review]
Awesome! I'm all for getting closer to the web spec, since there's clearly a lot of people that have thought longer about the problem than I ever have, nor will! And thanks for PR-ing it, it's much easier to talk about it when seeing the code!

Only the `PointerLeft.position` definition is a blocking concern for me (the macOS impl now doesn't emit `PointerLeft` because of it), the rest do not need to be resolved for this to still be an improvement.

File: src/event.rs [hunk 7]
```diff
@@ -896,12 +987,6 @@ pub enum Force {
         /// The value of this field is sufficiently high to provide a wide
         /// dynamic range for values of the `force` field.
         max_possible_force: f64,
-        /// The altitude (in radians) of the stylus.
-        ///
-        /// A value of 0 radians indicates that the stylus is parallel to the
-        /// surface. The value of this property is Pi/2 when the stylus is
-        /// perpendicular to the surface.
-        altitude_angle: Option<f64>,
```
Comment (madsmtm) [inline src/event.rs:904]: Is there a reason that this got removed?

File: src/event.rs (hunk 2 above)
Comment (daxpedda) [reply src/event.rs:268]: Will report back after some testing how other backends behave exactly, but would go with option 1.

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:493]: We have attempted to resolve this issue in *length* the other meeting, the outcome was not satisfactory, we should bring it up next meeting as well maybe together we can come up with a better solution.

In the previous proposal we had indeed `TouchCancel`, because it only applies to touch.
It might be a good idea indeed to implement `PointerCancel` and actually use it for other pointer types apart from touch?

This is definitely a weakness of this proposal, if we can't address it in this PR I will open an issue for it.

File: src/platform_impl/apple/uikit/view.rs [hunk 8]
```diff
@@ -454,25 +457,18 @@ impl WinitView {
         for touch in touches {
             let logical_location = touch.locationInView(None);
             let touch_type = touch.r#type();
-            let force = if os_supports_force {
+            let force = if let UITouchType::Pencil = touch_type {
+                None
```
Comment (madsmtm) [inline src/platform_impl/apple/uikit/view.rs:461]: Hmm, I think this is wrong? `UITouchType::Pencil` is exactly the kind of device that _does_ support force - even if we end up not using it below, I'd still rather compute the force, because we will need it later when we add proper support for pencils (and then this kind of branch is very confusing to see).

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:493]: Apologies, summarizing the context:

The reason why we couldn't come up with a good idea/didn't like `TouchCancel` is because it actually goes against the idea of the unified pointer model, where users don't need special handling to correctly handle any pointer type.

So folding it into `PointerLeft` without `PointerButton` with `Released` would make this "natural".
However for users who actually want to handle touch evens explicitly, its harder to make out `TouchCancel`.

So a `PointerCancel` event makes much more sense, because it solves this issue.
But I don't know how we can use it for anything else apart from touch.
However, that said, we can just use it for touch only, as long as users handle it regardless of pointer type, it should fulfill the original purpose.

I like it so far.

File: src/event.rs (hunk 4 above)
Comment (daxpedda) [reply src/event.rs:458]: > Splitting this from `PointerSource` does not feel sufficiently motivated, but perhaps it will make sense in the future because we're going to get more data on `PointerSource`?

That was the reason indeed.

> If we do keep it, then a naming suggestion could be to use `kind` and `PointerKind` instead of `ty` and `PointerType`.

Yeah, this is way better indeed.
Will adapt it!

File: src/event.rs (hunk 6 above)
Comment (daxpedda) [reply src/event.rs:303]: Yes, its needed for future pointer kinds.
Its also important that users can handle any pointer type, this is why `ButtonSource::mouse_button()` exists.
If we change it to `Option`, users will think no button was pressed on other pointer types, which would be incorrect. E.g. touch does indeed mimic the `MouseButton::Left` and so does regular pen input.

> I _feel_ like I'd rather mirror web even closer here, [..]

Actually the current API is how the Web API does this.
For touch and pen the same ID as the left mouse button is used, see the [W3C specification](https://w3c.github.io/pointerevents/#the-button-property).

File: src/event.rs (hunk 7 above)
Comment (daxpedda) [reply src/event.rs:904]: Yes, because its pen input and it doesn't exist in this current API.

My plan was to re-introduce the same code back again in #3810.
But if you want to I'm happy to introduce pen input in this PR directly.

File: src/platform_impl/apple/uikit/view.rs (hunk 8 above)
Comment (daxpedda) [reply src/platform_impl/apple/uikit/view.rs:461]: Same as https://github.com/rust-windowing/winit/pull/3876#discussion_r1721053489.

I can't use force here because I can't encode it into `PointerType/Source`, the `Unknown` variant can't hold `Force` and we can't just report it as `Touch`.

As mentioned above I'm happy to introduce `PointerType/Source::Pen` here just for MacOS and Windows.

File: src/event.rs (hunk 6 above)
Comment (madsmtm) [reply src/event.rs:303]: I'd feel even better with `button: MouseButton`!

But it's also fine to me if it makes sense for future pointer kinds (pens?), I'll trust your judgement on this (depending on how long you think it'll take before we get it implemented, if it'll take long, then I'd prefer the solution with fewer newtypes to start with).

File: src/event.rs (hunk 7 above)
Comment (madsmtm) [reply src/event.rs:904]: No, that's fine, I didn't read your PR description properly, that's on me ;)

File: src/event.rs (hunk 6 above)
Comment (daxpedda) [reply src/event.rs:303]: I agree!

My plan is to do it right after, if for some reason or another this doesn't happen, I promise to make a PR to address this!

File: src/platform_impl/apple/uikit/view.rs (hunk 8 above)
Comment (madsmtm) [reply src/platform_impl/apple/uikit/view.rs:461]: Yeah, I got that now, thanks!

File: src/event.rs (hunk 4 above)
Comment (daxpedda) [reply src/event.rs:458]: On another thought, this is also how Web works, another reason to switch to this new event.

File: src/event.rs (hunk 2 above)
Comment (daxpedda) [reply src/event.rs]: Apparently all backends report a position outside the window, which does make sense actually.

Fixed MacOS as well!

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:494]: On another thought, this is also how Web works, another reason to switch to this new event.

File: src/event.rs (hunk 4 above)
Comment (daxpedda) [reply src/event.rs]: Done!

File: src/event.rs (hunk 5 above)
Comment (daxpedda) [reply src/event.rs:873]: Done!

File: src/platform_impl/apple/appkit/view.rs [hunk 9]
```diff
@@ -639,19 +639,32 @@ declare_class!(
...
+        fn mouse_entered(&self, event: &NSEvent) {
             trace_scope!("mouseEntered:");
-            self.queue_event(WindowEvent::CursorEntered {
+
+            let view_point = self.mouse_view_point(event);
+
+            if !self.is_in_window(view_point) {
+                return
+            }
+
```
Comment (madsmtm) [inline src/platform_impl/apple/appkit/view.rs:650]: Nit: Redundant check (if the mouse just entered the view, then it cannot be outside of it).

```suggestion
```

File: src/platform_impl/apple/appkit/view.rs [hunk 10]
```diff
@@ -1034,42 +1047,63 @@ impl WinitView {
...
+    }
+
+    fn mouse_view_point(&self, event: &NSEvent) -> LogicalPosition<f64> {
         let window_point = unsafe { event.locationInWindow() };
         let view_point = self.convertPoint_fromView(window_point, None);
+
+        LogicalPosition::new(view_point.x, view_point.y)
+    }
+
+    fn is_in_window(&self, position: LogicalPosition<f64>) -> bool {
```
Comment (madsmtm) [inline src/platform_impl/apple/appkit/view.rs:1091]: Nit: Use AppKit terminology here.

```suggestion
    fn is_in_frame(&self, position: LogicalPosition<f64>) -> bool {
```

File: src/event.rs [hunk 11]
```diff
@@ -425,6 +449,130 @@ pub enum WindowEvent {
...
+    /// **macOS:** Unsupported.
+    Touch {
+        finger_id: FingerId,
+        force: Option<Force>,
+    },
+    Unknown(u16),
+}
+
+impl ButtonSource {
+    pub fn mouse_button(self) -> MouseButton {
```
Comment (madsmtm) [inline src/event.rs:554]: Nit: Missing docs

**Review by madsmtm — APPROVED** [review]
macOS and iOS impls look good now, thanks!

File: src/event.rs (hunk 3 above)
Comment (madsmtm) [reply src/event.rs:494]: I'm fine with leaving this to be figure out later

File: src/platform_impl/apple/appkit/view.rs (hunk 10 above)
Comment (daxpedda) [reply src/platform_impl/apple/appkit/view.rs]: So I did figure that this check didn't make sense in `mouse_click()` as well, so I removed this function entirely in https://github.com/rust-windowing/winit/pull/3876/commits/2aed0efa61b37424b481e7cc0609b3e35e7fd2ff.
Apologies for all the code sorting here, I just reverted back to the exact order it was before this PR so the squashed diff becomes as small as possible.

Let me know if I got anything wrong here!

Comment (madsmtm) [reply src/platform_impl/apple/appkit/view.rs]: Looks fine, thanks!

File: src/event.rs (hunk 11 above)
Comment (daxpedda) [reply src/event.rs:557]: Done!

File: src/platform_impl/android/mod.rs [hunk 12]
```diff
@@ -344,22 +344,58 @@ impl EventLoop {
...
-                            "Input event {device_id:?}, {phase:?}, loc={location:?}, \
+                            "Input event {device_id:?}, {phase:?}, loc={position:?}, \
                              pointer={pointer:?}"
                         );
+                        let finger_id = event::FingerId(FingerId(pointer.pointer_id()));
+                        let force = Some(Force::Normalized(pointer.pressure() as f64));
 
-                        let event = event::WindowEvent::Touch(event::Touch {
-                            device_id,
-                            phase,
```
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:356]: Now that we no longer output `phase`, does it make any sense to map to `TouchPhase`? This would only be used in the case of gestures which are not even implemented yet, perhaps we can remove this side-stepping?

File: src/platform_impl/android/mod.rs [hunk 13]
```diff
@@ -344,22 +344,58 @@ impl EventLoop {
...
+                                    device_id,
+                                    position,
+                                    kind: event::PointerKind::Touch(finger_id),
+                                };
+                                app.window_event(&self.window_target, window_id, event);
+                                let event = event::WindowEvent::PointerButton {
+                                    device_id,
+                                    state: event::ElementState::Pressed,
+                                    position,
+                                    button: event::ButtonSource::Touch { finger_id, force },
```
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:368]: I guess we should differentiate the source here as well? Android also supports pens and mouse input, and getting at least mouse input represented accurately seems like a sensible thing to do in this PR?

File: src/platform_impl/android/mod.rs (hunk 12 above)
Comment (daxpedda) [reply src/platform_impl/android/mod.rs:356]: Done!

File: src/platform_impl/android/mod.rs (hunk 13 above)
Comment (daxpedda) [reply src/platform_impl/android/mod.rs]: Mouse handling needs a bit more work, so I left it to a follow-up PR.
But I implemented differentiating the different tool types.
Entirely skipping the mouse tool type here shouldn't be a regression because before this PR it would show up as touch, which would have been wrong anyway.

Let me know what you think!

File: src/event.rs (hunk 3 above)
Comment (kchibisov) [reply src/event.rs:494]: I mean, your `PointerCancel` is `PointerLeave` because it means that you effectively lost the focus and thus shouldn't do anything in reaction to pointer input, it's the same for mouse/tablet and touch cancel being here makes sense, at least to me, since I was the one suggesting to do it that way.

Maybe it's not clear what `PointerLeave` actually does? Since what it says that you _lost focus_ it's not like you left the surface since it won't be emitted in case you have on-going grabs, so it's effectively a _cancel_ of the events, just with position attached.

If we add a generic `PointerCancel` it won't be any different than `PointerLeave` because that's what it does.

File: src/event.rs [hunk 14]
```diff
@@ -425,6 +449,133 @@ pub enum WindowEvent {
...
+    /// See [`PointerSource::Touch`] for more details.
+    ///
+    /// ## Platform-specific
+    ///
+    /// **macOS:** Unsupported.
+    Touch {
+        finger_id: FingerId,
+        force: Option<Force>,
+    },
+    Unknown(u16),
```
Comment (kchibisov) [inline src/event.rs:550]: What this `Unknown` is reserved for and why it's limited to u16?

File: src/event.rs [hunk 15]
```diff
@@ -1071,19 +1167,41 @@ mod tests {
                 with_window_event(HoveredFile("x.txt".into()));
                 with_window_event(HoveredFileCancelled);
                 with_window_event(Ime(Enabled));
-                with_window_event(CursorMoved { device_id: did, position: (0, 0).into() });
+                with_window_event(PointerMoved {
+                    device_id: did,
```
Comment (kchibisov) [inline src/event.rs:1171]: I start to wonder whether device ID should belong here in general, since what kind of device id Touch input has? Maybe it's a property of a specific input device, like `Mouse(DeviceId)`?

File: src/platform_impl/linux/wayland/seat/touch/mod.rs [hunk 16]
```diff
@@ -42,18 +42,26 @@ impl TouchHandler for WinitState {
...
-                    FingerId(id),
-                )),
-            }),
+            WindowEvent::PointerEntered {
+                device_id,
+                position,
+                kind: PointerKind::Touch(finger_id),
+            },
+            window_id,
+        );
```
Comment (kchibisov) [inline src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: Weren't we discussing that this should be emitted only for the first any finger touching the surface, so it'll look more like a pointer focus stuff?

File: src/platform_impl/linux/wayland/seat/touch/mod.rs [hunk 17]
```diff
@@ -87,18 +95,26 @@ impl TouchHandler for WinitState {
...
+                button: ButtonSource::Touch { finger_id, force: None },
+            },
+            window_id,
+        );
+        self.events_sink.push_window_event(
+            WindowEvent::PointerLeft {
+                device_id,
+                position: Some(position),
+                kind: PointerKind::Touch(finger_id),
+            },
```
Comment (kchibisov) [inline src/platform_impl/linux/wayland/seat/touch/mod.rs:117]: The same as _enter_ but for the last to leave finger.

**Review by kchibisov — CHANGES_REQUESTED** [review]
What I've commented on Wayland should apply to the rest, but in general it looks ok, I'd say.

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:494]: `PointerLeave` has nothing to do with focus.
While it is true that if _pointer focus_ is in play you won't get a `PointerLeave` until you release the pointer focus, this doesn't mean that any other scenario is not relevant.

`PointerLeave` means the pointer has left the surface, with the exception of pointer focus, in which case you will get `PointerLeave` upon loosing pointer focus if the pointer is not inside the surface.
`PointerCancel` means that _input_ should be discarded.

While in the case of touch input, `PointerLeave` and `PointerCancel` are the same thing, as @madsmtm pointed out, the problem is that it is currently hard to figure out when that actually happens because it depends on _not receiving_ a `PointerButton` with `Released`.
While it is possible that users just assume that `PointerLeave` means both and so they can handle it the same, it is not the same as pointed out above. Users who need to distinguish between these two scenarios will have a harder time doing so without a dedicated event.

I think the discussion should rather be around if this differentiation is actually needed, personally I don't know about that. On the other hand W3C deemed it necessary, so maybe digging there could yield something useful.

File: src/event.rs (hunk 14 above)
Comment (daxpedda) [reply src/event.rs:550]: It means its an unknown pointer type, the `u16` is the button ID, ergo the same as in `MouseButton::Other`.

File: src/event.rs (hunk 15 above)
Comment (daxpedda) [reply src/event.rs:1171]: I know at least one use-case: if your drawing tablet has touch support and your screen as well, these would be two different devices.

File: src/platform_impl/linux/wayland/seat/touch/mod.rs (hunk 16 above)
Comment (daxpedda) [reply src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: Don't remember :cold_sweat:.

I'm not sure what the right move is here, but I would argue that this makes more sense because it is simply more accurate. Unsure how not emitting `PointerEntered` would make it look more like pointer focus.

Currently I'm in favor of leaving it as-is because its accurate and I'm not sure what the advantage would be of changing it.

File: src/event.rs (hunk 3 above)
Comment (kchibisov) [reply src/event.rs:494]: >PointerLeave has nothing to do with focus.

pointer focus, sorry.

>PointerLeave means the pointer has left the surface, with the exception of pointer focus, in which case you will get PointerLeave upon loosing pointer focus if the pointer is not inside the surface.
PointerCancel means that input should be discarded.

This is not the case for grabs, if you want to do it that way, then you'd need to update all the implementations to account for that, which I don't think worth it, since you generally want a _pointer focus_, which what it models right now, at least on X11/Wayland.

>While it is possible that users just assume that PointerLeave means both and so they can handle it the same, it is not the same as pointed out above. Users who need to distinguish between these two scenarios will have a harder time doing so without a dedicated event.

I'll just say that you can get leave on regular mouse devices without releasing the button meaning that you should cancel out everything.

In general, I don't mind having a _cancel_ since I have a case on wayland where you can _remove the pointer_ all together meaning that you likely have to cancel everything/same for keyboard/touch/etc.

File: src/platform_impl/linux/wayland/seat/touch/mod.rs (hunk 16 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: In the current state of things there's no need for `PointerButton` for touch at all, since you can not touch without enter/leave meaning that enter/leave already model the touch down/up.

It's just events are redundant, since they end up doing the exact same thing, indicating touch down/up...

From what I can tell, maybe on some platforms you want to react when you get pointer focus/leave focus, but with touch you'll run this state change for each finger, which doesn't look _right_. Imagine decorations lightening, etc.

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:494]: > This is not the case for grabs, if you want to do it that way, then you'd need to update all the implementations to account for that, which I don't think worth it, since you generally want a _pointer focus_, which what it models right now, at least on X11/Wayland.

The way I described is how it works currently for Web and X11.
I assume this is how it works for every other backend as well already.
If not we should figure out how it works currently and how we want it to work!

What do you mean with "grab"? Don't you mean "pointer focus" with that?

File: src/platform_impl/linux/wayland/seat/touch/mod.rs (hunk 16 above)
Comment (daxpedda) [reply src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: > In the current state of things there's no need for `PointerButton` for touch at all, since you can not touch without enter/leave meaning that enter/leave already model the touch down/up.
> 
> It's just events are redundant, since they end up doing the exact same thing, indicating touch down/up...

Hm ... agreed!
Well, lets hash it out at the meeting today.
> From what I can tell, maybe on some platforms you want to react when you get pointer focus/leave focus, but with touch you'll run this state change for each finger, which doesn't look _right_. Imagine decorations lightening, etc.

Please elaborate.
Why would it be wrong to do "decorations lightening" on the second or third finger?
Wouldn't that also mean that you shouldn't do "decorations lightening" on `PointerButton`?

File: src/event.rs (hunk 3 above)
Comment (kchibisov) [reply src/event.rs:494]: I mean that if you press the button and move the pointer out of the window, you won't get the `leave` until you release the button, the coordinates client will be getting will be outside of the window, but the pointer is still _focusing_ the window, because you're holding the button.

So if you treat _leave_ as _left the surface area_ it means that you  should detect when the mouse is not over the window and sent event that it left, but I'd say that all backends prevent _pointer leave_ until you release everything...

File: src/platform_impl/linux/wayland/seat/touch/mod.rs (hunk 16 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: You react to `PointerEnter` not `PointerButton` and for mouse you can not get mulitple. With touch you already have decorations _lightened_ from the first touch input, the rest you should just ignore since user already interacting with the window, since you haven't got any _leave_ events.

Though, generally applications have it delivered via _separate_ event these dasy.

File: src/event.rs (hunk 3 above)
Comment (daxpedda) [reply src/event.rs:494]: Cool, we agree then on what `PointerLeave` means, seems this was a misunderstanding then!

File: src/platform_impl/linux/wayland/seat/touch/mod.rs (hunk 16 above)
Comment (daxpedda) [reply src/platform_impl/linux/wayland/seat/touch/mod.rs:57]: > You react to `PointerEnter` not `PointerButton` and for mouse you can not get mulitple.

The point is that users should specifically **not react** to `PointerButton` in this context.
This is questionable because its not obvious to a user why they should not.

> With touch you already have decorations _lightened_ from the first touch input, the rest you should just ignore since user already interacting with the window, since you haven't got any _leave_ events.

It might help if you explain what you mean with "decorations" here.
But the logic you are talking about here seems to build upon some rules that don't seem obvious to me, I don't know why users should react to this but not to that.

Maybe this is also something we should hash out in the meeting so you can explain this whole matter holistically.

**Review by kchibisov — APPROVED** [review]
I've rebased the PR, however I decided to not go for the `PointerEntered`/`PointerLeft` handling that I was suggesting. The main reason for that is that:

When you press a click with a pointer from e.g. unfocused window state, you generally have a sequence of `Enter -> Button`, and in generally, if you do button input, you have a preceding event of moving it in place. So having only a single `Enter`/`Leave` will
make button jump without any other event, which is a bit more inconsistent in my opinion that having more than one `PointerEntered`.

Fairly, there's nothing more with multiple `PointerEntered` if you account for multi-seat and similar devices.

--

I've rebased and squashed @daxpedda 's commits and added my small doc patch/changelog patch at the end.

@daxpedda could you look into what I've added, so if it's fine I'll merge this PR, since I have no desire to rebase it further.

File: src/event.rs [hunk 18]
```diff
@@ -455,12 +455,7 @@ pub enum WindowEvent {
     RedrawRequested,
 }
 
-/// Represents the pointer type of a pointer event.
-///
-/// ## Platform-specific
-///
-/// **Wayland/X11:** Can only identify [`Touch`](Self::Touch), any other pointer is identified as
-/// [`Mouse`](Self::Mouse).
```
Comment (daxpedda) [inline src/event.rs]: Why was this removed?
I believe this should still be valid, no?

**Review by daxpedda — COMMENTED** [review]
Just a small nit, otherwise LGTM.
Thank you @kchibisov for picking this up!

File: src/event.rs (hunk 18 above)
Comment (kchibisov) [reply src/event.rs]: It doesn't make any sense given that there's only `Mouse` and `Touch`, basically. And `Unknown` is more of a fallback here.

Comment (daxpedda) [reply src/event.rs]: I'm not following. The point is that unlike other backends, these backends can't differentiate between unknown and `Mouse`. E.g. while on every other backend a pen will identify as `Unknown`, on X11 and Wayland it identifies as `Mouse`, this is I believe an important discrepancy.

Comment (kchibisov) [reply src/event.rs]: Wayland can not identify pen because it's not implemented(I think the same for X11), so compositor sends it as _mouse_, that's true.

I can bring it back, but if something is emulated as mouse, it's a _mouse_, not any other device or Unknown. Like you don't have a case where something is `Unknown` and acts differently on them. 

I guess on other backends `Unknown` does act differently though.

Comment (daxpedda) [reply src/event.rs]: > Wayland can not identify pen because it's not implemented(I think the same for X11), so compositor sends it as _mouse_, that's true.

Well, the documentation described the current state, not any potential one.

> I can bring it back, but if something is emulated as mouse, it's a _mouse_, not any other device or Unknown. Like you don't have a case where something is `Unknown` and acts differently on them.
>
> I guess on other backends `Unknown` does act differently though.

What does that mean?
Why would it be *emulated* if its actually a mouse?
What do you mean with "acts differently"?

Comment (kchibisov) [reply src/event.rs]: Tablet input on Wayland right now arrives as mouse, because compositor emulates it, because winit doesn't support tablet input protocol. The same will happen with any other pointer input that could be emulated as mouse.

>What do you mean with "acts differently"

Unknown device isn't a mouse, it may not have buttons, etc. Like tablet(it doesn't have buttons, like on mouse).

Like the thing here is that not the issue with Wayland distinguishing devices, but with system ensures that client understands the input with the set of protocol it binds, so `Unknown` can not happed due to compositor making sure to convert everything to mouse.

Comment (daxpedda) [reply src/event.rs]: Okay, so you are saying that `Unknown` is never going to happen on Wayland.
It seems to me that the original documentation is correct and should remain in place then.

Comment (kchibisov) [reply src/event.rs]: I'll reword to `Unknown` can not happen, rather _can only identify touch_.

Comment (daxpedda) [reply src/event.rs]: Sounds good to me, lets sync up that change with all the other types as well please.

Comment (kchibisov) [reply src/event.rs]: Done.

Comment (daxpedda) [reply src/event.rs]: It should specifically say that its `Mouse` in this case.

Comment (kchibisov) [reply src/event.rs]: I'm not sure it matters that much though? Like if client on Wayland supports only Touch, it'll go only via touch.

Comment (daxpedda) [reply src/event.rs]: I would like our documentation to be accurate.
It seems very important to me that users are aware that pen events (or any other unsupported type) are emitted as `Mouse` events instead of `Unknown` on Wayland/X11.

I'm having a hard time understanding where you are coming from here and why you would like to not expose this kind of information to users.

Comment (kchibisov) [reply src/event.rs]: >It seems very important to me that users are aware that pen events (or any other unsupported type) are emitted as Mouse events instead of Unknown on Wayland/X11.

I want a bit more vague description of what happens because nothing is defined, if I say that everything is converted to `Mouse` then once some compositor will convert to `Touch`, it'll be _confusing_ as well (and I know compositor that converts even mouse to _touch_ for _reasons_).

```
/// **Wayland/X11:** [`Unknown`](Self::Unknown) can not occur due to system converting them to
/// known variants.
```

I can put something like that, since it's the most accurate thing I know.

Comment (daxpedda) [reply src/event.rs]: I see!

How about (just trying to improve the English, not the best myself):
```rust
/// **Wayland/X11:** [`Unknown`](Self::Unknown) device types are converted to known variants by the system.
```

Comment (kchibisov) [reply src/event.rs]: yeah, fine as well.
